app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/tagsselector/TagsSelectorPresenter.kt

```diff
@@ -112,20 +118,57 @@ class TagsSelectorPresenter(
         }
     }
 
+    private fun kindByExt(extension: String): String? {
+        return when (extension) {
+            in ImageMetaExtra.ACCEPTED_EXTENSIONS -> "IMAGE"
+            in VideoMetaExtra.ACCEPTED_EXTENSIONS -> "VIDEO"
+            in DocumentMetaExtra.ACCEPTED_EXTENSIONS -> "DOCUMENT"
+            else -> null
+        }
+    }
+
     fun calculateTagsAndSelection() {
         Log.d(TAGS_SELECTOR, "calculating tags and selection")
         if (storage == null || index == null)
             return
 
         val resources = index!!.listIds(prefix)
+
+        var myResources = mutableListOf<ResourceMeta>()
+        resources.map {
+            myResources.add(index!!.getMeta(it))
+        }
+
         val tagsByResources = storage!!.groupTagsByResources(resources)
-        val allTags = tagsByResources.values.flatten().toSet()
+
+        var tagsMap = mutableMapOf(1L to "".toSet() as Tags)
```

Comment:
This code is sketchy :) I think it's better to separate concept of "virtual tags" into other variable. There is no need to copy elements of `tagsByResources` manually and increment `resourceCounter` as well. Almost everything can be done with immutable methods like `length`, `map`, `filter` etc.

Reply:
Separate function to find the virtual tags

`    
tagsByResources = getVirtualTags(tagsByResources)
.
.
.

private fun getVirtualTags(
        tagsByResources: Map<ResourceId, Tags>
    ): Map<ResourceId, Tags> {
        var modifiedTagsByResources = tagsByResources
        if (showKindTags) {
            modifiedTagsByResources = modifiedTagsByResources.map { (id, tags) ->
                val listOfTags = tags.toMutableList()
                if (kindByExt(index!!.getMeta(id).extension) != null) {
                    listOfTags.add(KIND + kindByExt(index!!.getMeta(id).extension))
                }
                id to listOfTags.toSet()
            }.toMap()
        }
        return modifiedTagsByResources
    }`